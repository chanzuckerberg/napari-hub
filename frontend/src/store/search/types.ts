import { SearchResult } from './search.types';

/**
 * Function that transforms a list of search results into a different list of
 * search results after sorting or filtering.
 */
export type SearchResultTransformFunction = (
  results: SearchResult[],
) => SearchResult[];

/**
 * Response data from fetching the SPDX license list:
 * https://github.com/spdx/license-list-data/blob/master/json/licenses.json
 */
export interface SpdxLicenseResponse {
  licenseListVersion: string;
  licenses: SpdxLicenseData[];
  releaseDate: string;
}

/**
 * Response data for an SPDX license object.
 */
export interface SpdxLicenseData {
  reference: string;
  isDeprecatedLicenseId: boolean;
  detailsUrl: string;
  referenceNumber: number;
  name: string;
  licenseId: string;
  seeAlso: string[];
  isOsiApproved: boolean;
  isFsfLibre?: boolean;
}

export interface Resettable {
  reset(): void;
}
