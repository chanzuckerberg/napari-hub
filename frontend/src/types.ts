import { RequestError } from '@octokit/types';

export interface PluginAuthor {
  name: string;
  email?: string;
}

/**
 * Citation formats.
 * */
export interface CitationData {
  citation: string;
  RIS: string;
  BibTex: string;
  APA: string;
}

export type HubDimension =
  | 'Workflow step'
  | 'Supported data'
  | 'Image modality';

export type PluginCategory = Partial<{
  [key in HubDimension]: string[];
}>;

export type PluginCategoryHierarchy = Partial<{
  [key in HubDimension]: string[][];
}>;

/**
 * Plugin data used for indexing. This is a subset of the full plugin data.
 */
export interface PluginIndexData {
  authors: PluginAuthor[] | '';
  category?: PluginCategory | '';
  description_content_type: string;
  description: string;
  description_text: string;
  development_status: string[] | '';
  first_released: string;
  license: string;
  name: string;
  operating_system: string[] | '';
  python_version: string;
  release_date: string;
  summary: string;
  version: string;
}

/**
 * Interface for plugin data response from backend.
 */
export interface PluginData extends PluginIndexData {
  category_hierarchy?: PluginCategoryHierarchy;
  citations?: CitationData | null;
  code_repository: string;
  documentation: string;
  project_site: string;
  release_date: string;
  report_issues: string;
  requirements: string[] | '';
  support: string;
  twitter: string;
}

/**
 * Plugin repo data to render with plugin metadata.
 */
export interface PluginRepoData {
  forks: number;
  issuesAndPRs: number;
  stars: number;
}

export type PluginRepoFetchError = Pick<RequestError, 'name' | 'status'>;

/**
 * Data used for rendering links in the app.
 */
export interface LinkInfo {
  /**
   * URL of this link.
   */
  link: string;
  /**
   * Title of the link to use.
   */
  title: string;
  /**
   * If the link should open in a new tab.
   */
  newTab?: boolean;
}

// eslint-disable-next-line @typescript-eslint/ban-types
export type DeriveGet = <T extends object>(proxyObject: T) => T;
