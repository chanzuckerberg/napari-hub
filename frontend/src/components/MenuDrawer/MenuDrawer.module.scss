.item:not(:last-child) {
  @apply mb-5;
}
