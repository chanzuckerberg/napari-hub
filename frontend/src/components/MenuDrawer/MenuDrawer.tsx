import { IconButton, SwipeableDrawer } from '@material-ui/core';

import { Link } from '@/components/common';
import { Close } from '@/components/common/icons';
import { LinkInfo } from '@/types';

interface Props {
  items: LinkInfo[];
  onClose: () => void;
  onOpen: () => void;
  visible: boolean;
}

/**
 * Navigation drawer that slides out from the right. The drawer can be opened by
 * pressing the menu button or by swiping left from the right side of the
 * screen. Conversely, the drawer can be closed by clicking the close button,
 * swiping the drawer right, or clicking outside of the drawer area.
 */
export function MenuDrawer({ items, onOpen, onClose, visible }: Props) {
  return (
    <SwipeableDrawer
      anchor="right"
      classes={{ paper: 'bg-black flex-row w-9/12 p-6' }}
      onClose={onClose}
      onOpen={onOpen}
      open={visible}
      data-testid="menu"
    >
      <ul className="text-white flex-grow">
        {items.map((item) => (
          <li
            className="mb-5 last:m-0"
            key={item.title}
            data-testid="drawerItem"
          >
            <Link href={item.link} onClick={onClose}>
              {item.title}
            </Link>
          </li>
        ))}
      </ul>

      <IconButton
        className="self-start p-0"
        data-testid="drawerClose"
        onClick={onClose}
      >
        <Close />
      </IconButton>
    </SwipeableDrawer>
  );
}
