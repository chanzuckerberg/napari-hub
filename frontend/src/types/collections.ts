import { PluginAuthor, PluginData } from 'src/types';

export interface CollectionInstitution {
  institution: string;
  website: string;
}

export interface CollectionLinks {
  orcid: string;
  twitter?: string;
  github?: string;
  website?: string;
}

export interface CollectionCurator {
  name: string;
  title: string;
  affiliation: CollectionInstitution;
  links?: CollectionLinks;
}

export interface CollectionPlugin
  extends Pick<PluginData, 'name' | 'summary' | 'display_name'> {
  authors: PluginAuthor[];
  comment?: string;
}

export interface CollectionIndexData {
  title: string;
  cover_image: string;
  summary: string;
  curator: CollectionCurator;
  symbol: string;
}

export interface CollectionData extends CollectionIndexData {
  description: string;
  updated_date: Date;
  visible: boolean;
  plugins: CollectionPlugin[];
  visibility?: 'public' | 'hidden' | 'disabled';
}
