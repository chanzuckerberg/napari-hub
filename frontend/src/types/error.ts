import { RequestError } from '@octokit/types';

export type PluginRepoFetchError = Pick<RequestError, 'name' | 'status'>;
