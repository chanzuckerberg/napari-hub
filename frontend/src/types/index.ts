import { RequestError } from '@octokit/types';
import { NonFunctionKeys } from 'utility-types';

export interface PluginAuthor {
  name: string;
  email?: string;
}

/**
 * Citation formats.
 * */
export interface CitationData {
  citation: string;
  RIS: string;
  BibTex: string;
  APA: string;
}

export type HubDimension =
  | 'Workflow step'
  | 'Supported data'
  | 'Image modality';

export type PluginCategory = Partial<{
  [key in HubDimension]: string[];
}>;

export type PluginCategoryHierarchy = Partial<{
  [key in HubDimension]: string[][];
}>;

/**
 * Plugin data used for indexing. This is a subset of the full plugin data.
 */
export interface PluginIndexData {
  authors: PluginAuthor[] | '';
  category?: PluginCategory | '';
  description_content_type: string;
  description_text: string;
  description: string;
  development_status: string[] | '';
  display_name?: string;
  first_released: string;
  license: string;
  name: string;
  operating_system: string[] | '';
  python_version: string;
  release_date: string;
  summary: string;
  version: string;
  plugin_types?: PluginType[];
  reader_file_extensions?: string[];
  writer_file_extensions?: string[];
  writer_save_layers?: PluginWriterSaveLayer[];
}

/**
 * Interface for plugin data response from backend.
 */
export interface PluginData extends PluginIndexData {
  action_repository?: string;
  category_hierarchy?: PluginCategoryHierarchy;
  citations?: CitationData | null;
  code_repository: string;
  documentation: string;
  project_site: string;
  release_date: string;
  report_issues: string;
  requirements: string[] | '';
  support: string;
  twitter: string;
}

/**
 * Plugin repo data to render with plugin metadata.
 */
export interface PluginRepoData {
  forks: number;
  issuesAndPRs: number;
  stars: number;
}

export type PluginRepoFetchError = Pick<RequestError, 'name' | 'status'>;

/**
 * Data used for rendering links in the app.
 */
export interface LinkInfo {
  /**
   * URL of this link.
   */
  link: string;
  /**
   * Title of the link to use.
   */
  title: string;
  /**
   * If the link should open in a new tab.
   */
  newTab?: boolean;
}

export type ClassState<T extends object> = {
  [key in NonFunctionKeys<T>]: T[key];
};

// IMPORTANT: The order of the enum properties here matter. See:
// https://github.com/chanzuckerberg/napari-hub/pull/472#pullrequestreview-952464854
export enum PluginType {
  Widget = 'widget',
  Reader = 'reader',
  Writer = 'writer',
  SampleData = 'sample_data',
  Theme = 'theme',
}

export enum PluginWriterSaveLayer {
  Image = 'image',
  Points = 'points',
  Shapes = 'shapes',
}
