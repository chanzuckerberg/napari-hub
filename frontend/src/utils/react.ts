import { ReactElement, ReactNode } from 'react';

/**
 * Checks if a React node is an element. An element is an instantiated
 * component.
 *
 * @param node The React node
 * @returns True if node is an element
 */
export function isReactElement(node: ReactNode): node is ReactElement {
  const element = node as ReactElement | null | undefined;
  return !!(element?.type && element?.props);
}
