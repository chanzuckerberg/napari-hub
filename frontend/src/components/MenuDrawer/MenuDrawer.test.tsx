import { fireEvent, render, screen } from '@testing-library/react';
import { ComponentProps } from 'react';

import { LinkInfo } from '@/types';

import { MenuDrawer } from './MenuDrawer';

const noop = () => {};

describe('<MenuDrawer />', () => {
  const mockItems: LinkInfo[] = [
    {
      title: 'foo',
      link: '/foo',
    },
    {
      title: 'bar',
      link: '/bar',
    },
  ];

  type Props = ComponentProps<typeof MenuDrawer>;

  const renderComponent = ({
    items = mockItems,
    onClose = noop,
    onOpen = noop,
    visible = true,
  }: Partial<Props> = {}) => {
    const props = {
      items,
      onClose,
      onOpen,
      visible,
    } as Props;

    return render(<MenuDrawer {...props} />);
  };

  it('should match snapshot', () => {
    renderComponent({ visible: true });
    const menu = screen.getByTestId('menu');
    expect(menu).toMatchSnapshot();
  });

  it('should render items', () => {
    renderComponent({ visible: true });
    const items = screen.getAllByTestId('drawerItem');

    items.forEach((item, index) => {
      const link = item.querySelector('a');
      expect(link).toHaveTextContent(mockItems[index].title);
      expect(link).toHaveAttribute('href', mockItems[index].link);
    });
  });

  it('should call onClose when close button is clicked', () => {
    const onClose = jest.fn();
    renderComponent({ onClose, visible: true });

    const closeButton = screen.getByTestId('drawerClose');
    fireEvent.click(closeButton);

    expect(onClose).toHaveBeenCalled();
  });
});
