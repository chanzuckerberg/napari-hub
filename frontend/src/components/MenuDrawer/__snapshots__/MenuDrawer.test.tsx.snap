// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuDrawer /> should match snapshot 1`] = `
<div
  class="MuiDrawer-root MuiDrawer-modal"
  data-testid="menu"
  role="presentation"
  style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  />
  <div
    data-test="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiPaper-root MuiDrawer-paper bg-black flex-row w-9/12 p-6 MuiDrawer-paperAnchorRight MuiPaper-elevation16"
    tabindex="-1"
  >
    <ul
      class="text-white flex-grow"
    >
      <li
        class="mb-5 last:m-0"
        data-testid="drawerItem"
      >
        <a
          href="/foo"
        >
          foo
        </a>
      </li>
      <li
        class="mb-5 last:m-0"
        data-testid="drawerItem"
      >
        <a
          href="/bar"
        >
          bar
        </a>
      </li>
    </ul>
    <button
      class="MuiButtonBase-root MuiIconButton-root self-start p-0"
      data-testid="drawerClose"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          fill="none"
          height="18"
          viewBox="0 0 18 18"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1 1L17 17"
            stroke="white"
            stroke-width="2.35294"
          />
          <path
            d="M17 1L1 17"
            stroke="white"
            stroke-width="2.35294"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    data-test="sentinelEnd"
    tabindex="0"
  />
</div>
`;
